homeassistant/components/p1_monitor/config_flow.py

```diff
@@ -51,6 +52,7 @@ async def async_step_user(
                         CONF_NAME, default=self.hass.config.location_name
                     ): str,
                     vol.Required(CONF_HOST): str,
+                    vol.Optional(CONF_WATERMETER, default=False): bool,
```

Comment:
Why does an user need to configure this? Can't we just add (or not add) this based on the data we get/is available?

Reply:
Seems like the better way to me, but I couldn't figure it out in the code.

Reply:
You could just call `self.p1monitor.watermeter()` and see if it works or not? And decide in the code based on that result?

Reply:
> You could just call `self.p1monitor.watermeter()` and see if it works or not? And decide in the code based on that result?

Is it necessary that the user doesn't need any interaction? I think most of the people don't have a watermeter combined (it's quite a hassle with extra hardware), so default no seems appropriate. It's not as easy as reading the P1 results.

I hope this change can be approved somehow. I am really looking forward to this feature.

Reply:
Did some testing and was able to make some improvements, will push the code later.

Reply:
Code has been changed, the config_flow option is no longer necessary, however the way I've done it now it will continuously throw a debug line in the logger if you don't use a water meter.

That could possibly be better, but didn't get a working if statement at the moment. The current situation is that you get a NoData exception when you have not set a watermeter in P1 Monitor.

Reply:
```diff
diff --git a/homeassistant/components/p1_monitor/__init__.py b/homeassistant/components/p1_monitor/__init__.py
index 3f8bbc5608..9ee3206213 100644
--- a/homeassistant/components/p1_monitor/__init__.py
+++ b/homeassistant/components/p1_monitor/__init__.py
@@ -70,6 +70,7 @@ class P1MonitorDataUpdateCoordinator(DataUpdateCoordinator[P1MonitorData]):
     """Class to manage fetching P1 Monitor data from single endpoint."""
 
     config_entry: ConfigEntry
+    has_water_meter: bool | None = None
 
     def __init__(
         self,
@@ -95,9 +96,13 @@ class P1MonitorDataUpdateCoordinator(DataUpdateCoordinator[P1MonitorData]):
             SERVICE_SETTINGS: await self.p1monitor.settings(),
             SERVICE_WATERMETER: None,
         }
-        try:
-            data[SERVICE_WATERMETER] = await self.p1monitor.watermeter()
-        except P1MonitorNoDataError:
-            LOGGER.debug("No watermeter data received from P1 Monitor")
+
+        if self.has_water_meter or self.has_water_meter is None:
+            try:
+                data[SERVICE_WATERMETER] = await self.p1monitor.watermeter()
+            except P1MonitorNoDataError:
+                LOGGER.debug("No watermeter data received from P1 Monitor")
+                if self.has_water_meter is None:
+                    self.has_water_meter = False
 
         return data
```

How about something like this?

Reply:
Exactly what was still missing, the only thing that still needs to be adjusted are the tests. Then I think this PR is finished.

Reply:
Tests and fixes

```diff
diff --git a/homeassistant/components/p1_monitor/__init__.py b/homeassistant/components/p1_monitor/__init__.py
index 9ee3206213..b157f3e811 100644
--- a/homeassistant/components/p1_monitor/__init__.py
+++ b/homeassistant/components/p1_monitor/__init__.py
@@ -100,6 +100,7 @@ class P1MonitorDataUpdateCoordinator(DataUpdateCoordinator[P1MonitorData]):
         if self.has_water_meter or self.has_water_meter is None:
             try:
                 data[SERVICE_WATERMETER] = await self.p1monitor.watermeter()
+                self.has_water_meter = True
             except P1MonitorNoDataError:
                 LOGGER.debug("No watermeter data received from P1 Monitor")
                 if self.has_water_meter is None:
diff --git a/homeassistant/components/p1_monitor/sensor.py b/homeassistant/components/p1_monitor/sensor.py
index 4053c86405..757bf249ca 100644
--- a/homeassistant/components/p1_monitor/sensor.py
+++ b/homeassistant/components/p1_monitor/sensor.py
@@ -240,7 +240,7 @@ async def async_setup_entry(
     hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
 ) -> None:
     """Set up P1 Monitor Sensors based on a config entry."""
-    coordinator = hass.data[DOMAIN][entry.entry_id]
+    coordinator: P1MonitorDataUpdateCoordinator = hass.data[DOMAIN][entry.entry_id]
     entities: list[P1MonitorSensorEntity] = []
     entities.extend(
         P1MonitorSensorEntity(
@@ -272,7 +272,7 @@ async def async_setup_entry(
         )
         for description in SENSORS_SETTINGS
     )
-    if coordinator.data[SERVICE_WATERMETER]:
+    if coordinator.has_water_meter:
         entities.extend(
             P1MonitorSensorEntity(
                 coordinator=coordinator,
diff --git a/tests/components/p1_monitor/conftest.py b/tests/components/p1_monitor/conftest.py
index 95e413340e..5f6713f522 100644
--- a/tests/components/p1_monitor/conftest.py
+++ b/tests/components/p1_monitor/conftest.py
@@ -48,7 +48,7 @@ def mock_p1monitor():
                 json.loads(load_fixture("p1_monitor/watermeter.json"))
             )
         )
-        yield p1monitor_mock
+        yield client
 
 
 @pytest.fixture
diff --git a/tests/components/p1_monitor/test_sensor.py b/tests/components/p1_monitor/test_sensor.py
index c15fa0637b..12060509ab 100644
--- a/tests/components/p1_monitor/test_sensor.py
+++ b/tests/components/p1_monitor/test_sensor.py
@@ -1,6 +1,9 @@
 """Tests for the sensors provided by the P1 Monitor integration."""
+from unittest.mock import MagicMock
 import pytest
 
+from p1monitor import P1MonitorNoDataError
+
 from homeassistant.components.p1_monitor.const import DOMAIN
 from homeassistant.components.sensor import (
     ATTR_STATE_CLASS,
@@ -213,6 +216,20 @@ async def test_watermeter(
     assert not device_entry.sw_version
 
 
+async def test_no_watermeter(
+    hass: HomeAssistant, mock_p1monitor: MagicMock, mock_config_entry: MockConfigEntry
+) -> None:
+    """Test the P1 Monitor - Without WaterMeter sensors."""
+    mock_p1monitor.watermeter.side_effect = P1MonitorNoDataError
+    mock_config_entry.add_to_hass(hass)
+    await hass.config_entries.async_setup(mock_config_entry.entry_id)
+    await hass.async_block_till_done()
+
+    assert not hass.states.get("sensor.watermeter_consumption_day")
+    assert not hass.states.get("sensor.consumption_total")
+    assert not hass.states.get("sensor.pulse_count")
+
+
 @pytest.mark.parametrize(
     "entity_id",
     ("sensor.smartmeter_gas_consumption",),
```

Reply:
PR is now fully updated